Add 'Span::parent()' and 'Span::source()' to proc_macro API.

SergioBenitez opened the pull request:
As the title suggests: a couple of useful methods for `proc_macro`.

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
r? @jseyfried

kennytm commented on src/libproc_macro/lib.rs:
```diff
@@ -221,6 +221,19 @@ impl Span {
 
+    /// The `Span` for the tokens in the previous macro expansion from which
+    /// `self` was generated from, if any.
+    pub fn parent(&self) -> Option<Span> {
```
Needs `#[unstable(feature = "proc_macro")]`. (Similar for `source()`)

```
[00:07:40] error: This node does not have a stability attribute
[00:07:40]    --> /checkout/src/libproc_macro/lib.rs:226:5
[00:07:40]     |
[00:07:40] 226 | /     pub fn parent(&self) -> Option<Span> {
[00:07:40] 227 | |         self.0.ctxt().outer().expn_info().map(|i| Span(i.call_site))
[00:07:40] 228 | |     }
[00:07:40]     | |_____^
[00:07:40] 
[00:07:40] error: This node does not have a stability attribute
[00:07:40]    --> /checkout/src/libproc_macro/lib.rs:233:5
[00:07:40]     |
[00:07:40] 233 | /     pub fn source(&self) -> Span {
[00:07:40] 234 | |         Span(self.0.source_callsite())
[00:07:40] 235 | |     }
[00:07:40]     | |_____^
```

SergioBenitez commented:
Yes, of course. I'll add some ui-tests as well.

jseyfried commented:
Great!
Alternative naming idea: `parent` -> `source`, `source` -> `original_source`. `parent` sounds a little strange to me w.r.t. spans. I don't have a super strong opinion though.

cc @nrc

SergioBenitez commented:
Added tests and fixed up the definition. I don't have a strong opinion about naming. `parent` simply felt more natural to me as I'm "moving up the chain".

SergioBenitez commented:
@jseyfried Look okay?

jseyfried commented:
Yeah, lets land. @bors r+

bors commented:
:pushpin: Commit ab365be has been approved by `jseyfried`

<!-- @bors r=jseyfried ab365be1406919d4da42442b70c153296980ba52 -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #47156

bors commented:
:pushpin: Commit ab365be has been approved by `jseyfried`

<!-- @bors r=jseyfried ab365be1406919d4da42442b70c153296980ba52 -->

bors commented:
:hourglass: Testing commit ab365be1406919d4da42442b70c153296980ba52 with merge a9a03d9bfb0e82322439fe6c252ebd4ba6f23d98...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5956), [status-travis](https://travis-ci.org/rust-lang/rust/builds/325763388?utm_source=github_status&utm_medium=notification)
Approved by: jseyfried
Pushing a9a03d9bfb0e82322439fe6c252ebd4ba6f23d98 to master...
